agent/worker.go

```diff
@@ -219,6 +229,57 @@ func reconcileTaskState(ctx context.Context, w *worker, added []*api.Task, remov
 	return tx.Commit()
 }
 
+// AssignSecrets assigns the set of secrets to the worker. Any secrets not in this set
+// will be removed.
+func (w *worker) AssignSecrets(ctx context.Context, secrets []*api.Secret) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	log.G(ctx).WithFields(logrus.Fields{
+		"len(secrets)": len(secrets),
+	}).Debug("(*worker).AssignSecrets")
+
+	return reconcileSecrets(ctx, w, secrets, nil, true)
+}
+
+// UpdateSecrets updates the set of secrets assigned to the worker.
+// Serets in the added set will be added to the worker, and secrets in the removed set
+// will be removed from the worker.
+func (w *worker) UpdateSecrets(ctx context.Context, added []*api.Secret, removed []string) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	log.G(ctx).WithFields(logrus.Fields{
+		"len(added)":   len(added),
+		"len(removed)": len(removed),
+	}).Debug("(*worker).UpdateSecrets")
+
+	return reconcileSecrets(ctx, w, added, removed, false)
+}
+
+func reconcileSecrets(ctx context.Context, w *worker, added []*api.Secret, removed []string, fullSnapshot bool) error {
+	// If this was a complete set of secrets, we're going to clear the secrets map and add all of them
+	w.secrets.RLock()
+	defer w.secrets.RUnlock()
+	if fullSnapshot {
+		w.secrets.m = make(map[string]*api.Secret)
+		for _, secret := range added {
+			w.secrets.m[secret.Spec.Annotations.Name] = secret
```

Comment:
Changed these to call CRD apis on secret instead, which calls `mu.Lock` instead when adding or removing or resetting.